envisage/plugins/ipython_kernel/tests/test_internal_ipkernel.py

```diff
@@ -105,29 +170,214 @@ def test_shutdown_closes_console_pipes(self):
+
     def test_io_pub_thread_stopped(self):
+        self.create_and_destroy_kernel()
+        io_pub_threads = self.objects_of_type(ipykernel.iostream.IOPubThread)
+        for thread in io_pub_threads:
+            self.assertFalse(thread.thread.is_alive())
+
+    def test_no_threads_leaked(self):
+        threads_before = threading.active_count()
+        self.create_and_destroy_kernel()
+        threads_after = threading.active_count()
+        self.assertEqual(threads_before, threads_after)
+
+    def test_no_new_atexit_handlers(self):
+        # Caution: this is a rather fragile and indirect test. We want
+        # to know that all cleanup has happened when shutting down the
+        # kernel, with none of that cleanup deferred to atexit handlers.
+        #
+        # Since we have no direct way to get hold of the atexit handlers on
+        # Python 3, we instead use the number of referents from the
+        # atexit module as a proxy.
+        #
+        # If this test starts failing, try adding a warmup cycle. If the
+        # first call to self.create_and_destroy_kernel adds new referents,
+        # that's not a big deal. But if every call consistently adds new
+        # referents, then there's something to be fixed.
+        atexit_handlers_before = len(gc.get_referents(atexit))
+        self.create_and_destroy_kernel()
+        atexit_handlers_after = len(gc.get_referents(atexit))
+
+        self.assertEqual(atexit_handlers_before, atexit_handlers_after)
+
+    def test_zmq_sockets_closed(self):
+        # Previously, tests were leaking file descriptors linked to
+        # zmq.Socket objects. Check that all extant sockets are closed.
+        self.create_and_destroy_kernel()
+        sockets = self.objects_of_type(zmq.Socket)
+        self.assertTrue(all(socket.closed for socket in sockets))
+
+    def test_ipykernel_live_objects(self):
+        # Check that all IPython-related objects have been cleaned up
+        # as expected.
+
+        self.create_and_destroy_kernel()
+
+        # Sometimes once isn't enough ....
+        # XXX Ideally, this wouldn't be necessary at all; see if we can fix.
+        for _ in range(3):
+            gc.collect()
+
+        shells = self.objects_of_type(ipykernel.zmqshell.ZMQInteractiveShell)
+        self.assertEqual(shells, [])
+
+        kernels = self.objects_of_type(ipykernel.ipkernel.IPythonKernel)
+        self.assertEqual(kernels, [])
+
+        kernel_apps = self.objects_of_type(ipykernel.kernelapp.IPKernelApp)
+        self.assertEqual(kernel_apps, [])
+
+    def test_ctrl_c_message_suppressed(self):
+        captured = []
+        with capture_stream(sys.__stdout__, callback=captured.append):
+            kernel = InternalIPKernel()
+            kernel.init_ipkernel(gui_backend=None)
+            kernel.shutdown()
+
+        self.assertEqual(len(captured), 1)
+        captured_stdout = captured[0].decode("utf-8")
+        self.assertNotIn("Ctrl-C will not work", captured_stdout)
```

Comment:
Just so I understand correcrltly, you're checking if a string doesn't exist within another string here? 

And can you comment on what `captured_stdout` is and maybe add a test for that?

Reply:
Sorry, to be clear, yes, `captured_stdout` is a (Unicode) string, representing everything that was written out by the kernel. And yes, the `assertNotIn` is doing a string-in-string test to check that the Ctrl-C message suppression is working.